File: src/tokamak/classify.py
"""5-factor error taxonomy for agent traces."""

from __future__ import annotations

import json
import re
from typing import Any, Dict

from .quality import _extract_messages, _content_string

ERROR_CLASSES = (
    "tool_failure",
    "syntax_error",
    "reasoning_error",
    "safety_refusal",
    "timeout_stall",
    "none",
)

_TOOL_FAIL = re.compile(
    r"(?i)\b(?:tool\s+(?:failed|error|timeout)|exit\s+code\s+[1-9]|"
    r"command\s+not\s+found|error\s*:\s*\w+|traceback)"
)
_SYNTAX = re.compile(
    r"(?i)\b(syntaxerror|parseerror|unexpected\s+token|invalid\s+syntax|"
    r"unmatched\s+(?:bracket|paren|brace))"
)
_REASONING = re.compile(
    r"(?i)\b(i\s+made\s+an?\s+(?:mistake|error)|that\s+was\s+wrong|"
    r"let\s+me\s+reconsider|i\s+was\s+(?:wrong|incorrect)|"
    r"actually,?\s+(?:that's|that\s+is)\s+(?:wrong|incorrect))"
)
_SAFETY = re.compile(
    r"(?i)\b(i\s+can'?t\s+(?:help|assist|do)|i'?m\s+(?:not\s+able|unable)|"
    r"i\s+(?:must|cannot)\s+decline|against\s+my\s+guidelines)"
)
_TIMEOUT = re.compile(
    r"(?i)\b(timeout|timed\s+out|deadline\s+exceeded|hung|stalled|"
    r"no\s+response|connection\s+reset)"
)


def classify(trace: Dict[str, Any]) -> str:
    text = json.dumps(trace, ensure_ascii=False)
    if _TOOL_FAIL.search(text):
        return "tool_failure"
    if _SYNTAX.search(text):
        return "syntax_error"
    if _SAFETY.search(text):
        return "safety_refusal"
    if _TIMEOUT.search(text):
        return "timeout_stall"
    if _REASONING.search(text):
        return "reasoning_error"
    return "none"

File: rust/tokamak/src/classify.rs
//! 5-factor error taxonomy.

use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::Value;

pub const ERROR_CLASSES: &[&str] = &[
    "tool_failure", "syntax_error", "reasoning_error",
    "safety_refusal", "timeout_stall", "none",
];

static TOOL_FAIL: Lazy<Regex> = Lazy::new(|| Regex::new(
    r"(?i)\b(?:tool\s+(?:failed|error|timeout)|exit\s+code\s+[1-9]|command\s+not\s+found|error\s*:\s*\w+|traceback)"
).unwrap());
static SYNTAX:    Lazy<Regex> = Lazy::new(|| Regex::new(
    r"(?i)\b(syntaxerror|parseerror|unexpected\s+token|invalid\s+syntax|unmatched\s+(?:bracket|paren|brace))"
).unwrap());
static REASONING: Lazy<Regex> = Lazy::new(|| Regex::new(
    r"(?i)\b(i\s+made\s+an?\s+(?:mistake|error)|that\s+was\s+wrong|let\s+me\s+reconsider|i\s+was\s+(?:wrong|incorrect)|actually,?\s+(?:that's|that\s+is)\s+(?:wrong|incorrect))"
).unwrap());
static SAFETY:    Lazy<Regex> = Lazy::new(|| Regex::new(
    r"(?i)\b(i\s+can'?t\s+(?:help|assist|do)|i'?m\s+(?:not\s+able|unable)|i\s+(?:must|cannot)\s+decline|against\s+my\s+guidelines)"
).unwrap());
static TIMEOUT:   Lazy<Regex> = Lazy::new(|| Regex::new(
    r"(?i)\b(timeout|timed\s+out|deadline\s+exceeded|hung|stalled|no\s+response|connection\s+reset)"
).unwrap());

pub fn classify(trace: &Value) -> &'static str {
    let text = serde_json::to_string(trace).unwrap_or_default();
    if TOOL_FAIL.is_match(&text)  { return "tool_failure"; }
    if SYNTAX.is_match(&text)     { return "syntax_error"; }
    if SAFETY.is_match(&text)     { return "safety_refusal"; }
    if TIMEOUT.is_match(&text)    { return "timeout_stall"; }
    if REASONING.is_match(&text)  { return "reasoning_error"; }
    "none"
}
